applications/CoSimulationApplication/co_simulation_api/co_sim_comm.h

```diff
@@ -0,0 +1,155 @@
+        { throw std::runtime_error("Type of data not yet supported"); }                          \
+    virtual bool ExportDetail(const DataContainerType& rDataContainer, const std::string& rIdentifier) \
+        { throw std::runtime_error("Type of data not yet supported"); }                          \
+
+class CoSimComm
+{
+public:
+    typedef Tools::SettingsType SettingsType;
+
+    explicit CoSimComm(const std::string& rName, SettingsType& rSettings, const bool IsConnectionMaster) : mrSettings(rSettings),mName(rName), mIsConnectionMaster(IsConnectionMaster)
+    {
+        const SettingsType default_settings = {
+            {"echo_level",   "1"},
+            {"print_timing", "0"}
+        };
+        Tools::AddMissingSettings(default_settings, mrSettings);
+
+        mEchoLevel = std::stoi(mrSettings.at("echo_level"));
+        mPrintTiming = (mrSettings.at("print_timing") == "1");
+    }
+
+    virtual ~CoSimComm() = default; // impl of disconnect has to be in derived class due to order of class destruction
+
+    bool Connect()
+    {
+        CS_LOG << "Connecting \"" << mName << "\" as Connection-" << (mIsConnectionMaster ? "MASTER" : "SLAVE") << " ..." << std::endl;
+
+        if (mIsConnected) {
+            throw std::runtime_error("A connection was already established!");
+        }
+
+        mIsConnected = ConnectDetail();
+
+        if (!mIsConnected) {
+            throw std::runtime_error("Connection was not successful!");
+        }
+
+        CS_LOG << "Connection established" << std::endl;
+
+        return mIsConnected;
+    }
+
+    bool Disconnect()
+    {
+        CS_LOG << "Disconnecting \"" << mName << "\" ..." << std::endl;
+
+        if (mIsConnected) {
+            mIsConnected = !DisconnectDetail();
+            if (mIsConnected) {
+                CS_LOG << "Warning: Disconnect was not successful!" << std::endl;
+                return false;
+            }
+        } else {
+            CS_LOG << "Warning: Calling Disconnect but there was no active connection!" << std::endl;
+            return false;
+        }
+
+        CS_LOG << "Disconnecting successful" << std::endl;
+
+        return true;
+    }
+
+    void SendControlSignal(const int rSignal, const std::string& rIdentifier)
+    {
+        CheckConnection(); return SendControlSignalDetail(rSignal, rIdentifier);
+    }
+    int RecvControlSignal(std::string& rIdentifier)
+    {
+        CheckConnection(); return RecvControlSignalDetail(rIdentifier);
+    }
+
+    CO_SIM_COMM_REGISTER_DATA_CONTAINER_TYPE(DataContainers::Geometry);
+    CO_SIM_COMM_REGISTER_DATA_CONTAINER_TYPE(DataContainers::Mesh);
+    CO_SIM_COMM_REGISTER_DATA_CONTAINER_TYPE(DataContainers::Data);
+
+protected:
+    SettingsType& mrSettings;
+
+    std::string GetName() const        {return mName;}
+    int GetEchoLevel() const           {return mEchoLevel;}
+    bool GetIsConnectionMaster() const {return mIsConnectionMaster;}
```

Comment:
one side (does not matter which one) should be the master to perform some operations.
e.g. for File-Communication this would be the creation / deletion of the folder in which the comm happens
for MPI-Comm one side would open ports, the other side would connect to it
=> the different sides do different calls